File: src/db-toolkit/operations/ai/schema_analyzer.py
"""Schema Analyzer for DBAssist AI functionality."""

from typing import Dict, List, Any
from .gemini_client import gemini_client


class SchemaAnalyzer:
    """AI assistant for database schema analysis."""

    def __init__(self):
        """Initialize schema analyzer."""
        self.client = gemini_client

    async def analyze_table(
        self, 
        table_name: str, 
        columns: List[Dict], 
        db_type: str = "postgresql"
    ) -> Dict[str, Any]:
        """Analyze a database table and provide insights."""
        try:
            result = await self.client.analyze_table(
                table_name=table_name,
                columns=columns,
                db_type=db_type
            )
            return {
                "success": True,
                "table_name": table_name,
                "summary": result.get("summary", ""),
                "purpose": "Data storage and retrieval",
                "suggestions": result.get("suggestions", []),
                "common_queries": result.get("common_queries", []),
                "index_recommendations": result.get("index_recommendations", []),
                "relationship_suggestions": []
            }
        except Exception as e:
            return {
                "success": False,
                "error": str(e),
                "table_name": table_name,
                "summary": "",
                "suggestions": []
            }

    async def suggest_indexes(
        self, 
        table_name: str, 
        columns: List[Dict],
        query_patterns: List[str] = None,
        db_type: str = "postgresql"
    ) -> Dict[str, Any]:
        """Suggest database indexes for better performance."""
        try:
            # Basic index suggestions based on column types and patterns
            suggestions = []
            
            for column in columns:
                col_name = column.get('name', '')
                col_type = column.get('type', '').lower()
                
                # Suggest indexes for common patterns
                if 'id' in col_name.lower() and col_name != 'id':
                    suggestions.append({
                        "type": "btree",
                        "columns": [col_name],
                        "reason": f"Foreign key column {col_name} would benefit from an index"
                    })
                elif col_type in ['timestamp', 'date', 'datetime']:
                    suggestions.append({
                        "type": "btree", 
                        "columns": [col_name],
                        "reason": f"Date/time column {col_name} commonly used in WHERE clauses"
                    })
                elif 'email' in col_name.lower():
                    suggestions.append({
                        "type": "btree",
                        "columns": [col_name],
                        "reason": f"Email column {col_name} likely used for lookups"
                    })

            return {
                "success": True,
                "table_name": table_name,
                "index_suggestions": suggestions,
                "performance_impact": "medium"
            }
        except Exception as e:
            return {
                "success": False,
                "error": str(e),
                "index_suggestions": []
            }

    async def analyze_relationships(
        self, 
        tables: Dict[str, List[Dict]], 
        db_type: str = "postgresql"
    ) -> Dict[str, Any]:
        """Analyze potential relationships between tables."""
        try:
            relationships = []
            table_names = list(tables.keys())
            
            # Look for potential foreign key relationships
            for table_name, columns in tables.items():
                for column in columns:
                    col_name = column.get('name', '').lower()
                    
                    # Check if column name suggests a foreign key
                    if col_name.endswith('_id') and col_name != 'id':
                        referenced_table = col_name[:-3]  # Remove '_id'
                        
                        # Check if referenced table exists
                        for other_table in table_names:
                            if other_table.lower() == referenced_table or other_table.lower() == referenced_table + 's':
                                relationships.append({
                                    "from_table": table_name,
                                    "from_column": column.get('name'),
                                    "to_table": other_table,
                                    "to_column": "id",
                                    "relationship_type": "many_to_one",
                                    "confidence": "medium"
                                })

            return {
                "success": True,
                "relationships": relationships,
                "total_suggestions": len(relationships)
            }
        except Exception as e:
            return {
                "success": False,
                "error": str(e),
                "relationships": []
            }

    async def generate_common_queries(
        self, 
        table_name: str, 
        columns: List[Dict],
        db_type: str = "postgresql"
    ) -> Dict[str, Any]:
        """Generate common SQL queries for a table."""
        try:
            queries = []
            col_names = [col.get('name') for col in columns if col.get('name')]
            
            # Basic SELECT query
            queries.append({
                "name": "Select All Records",
                "query": f"SELECT * FROM {table_name} LIMIT 10;",
                "description": "Retrieve first 10 records from the table"
            })
            
            # Count query
            queries.append({
                "name": "Count Records",
                "query": f"SELECT COUNT(*) FROM {table_name};",
                "description": "Count total number of records"
            })
            
            # Find queries based on column types
            for column in columns:
                col_name = column.get('name', '')
                col_type = column.get('type', '').lower()
                
                if 'timestamp' in col_type or 'date' in col_type:
                    queries.append({
                        "name": f"Recent Records by {col_name}",
                        "query": f"SELECT * FROM {table_name} WHERE {col_name} >= NOW() - INTERVAL '7 days' ORDER BY {col_name} DESC;",
                        "description": f"Find records from the last 7 days based on {col_name}"
                    })
                elif 'email' in col_name.lower():
                    queries.append({
                        "name": f"Find by {col_name}",
                        "query": f"SELECT * FROM {table_name} WHERE {col_name} = 'example@email.com';",
                        "description": f"Find records by {col_name}"
                    })

            return {
                "success": True,
                "table_name": table_name,
                "queries": queries[:5],  # Limit to 5 queries
                "total_generated": len(queries)
            }
        except Exception as e:
            return {
                "success": False,
                "error": str(e),
                "queries": []
            }


# Global instance
schema_analyzer = SchemaAnalyzer()

File: src/db-toolkit/electron/backend/operations/ai/schema-analyzer.js
/**
 * Schema Analyzer for DBAssist AI functionality.
 */

const { CloudflareAIClient } = require('./cloudflare-client');
const { SCHEMA_ANALYSIS_PROMPT, SYSTEM_PROMPT } = require('./prompts');

class SchemaAnalyzer {
  constructor(accountId, apiToken, model = '@cf/meta/llama-3.1-70b-instruct', temperature = 0.7, maxTokens = 2048) {
    if (!accountId || !apiToken) {
      throw new Error('Cloudflare credentials not configured');
    }

    this.client = new CloudflareAIClient(accountId, apiToken);
    this.model = model;
    this.temperature = temperature;
    this.maxTokens = maxTokens;
  }

  async analyzeTable(tableName, columns, dbType = 'postgresql') {
    try {
      let schemaInfo = `Table: ${tableName}\n`;
      schemaInfo += 'Columns:\n';
      for (const col of columns) {
        schemaInfo += `  - ${col.name} (${col.type || 'unknown'})\n`;
      }

      const prompt = SCHEMA_ANALYSIS_PROMPT
        .replace('{db_type}', dbType)
        .replace('{schema_info}', schemaInfo);

      const response = await this.client.generate(
        prompt,
        this.model,
        this.maxTokens,
        this.temperature,
        SYSTEM_PROMPT
      );

      return {
        success: true,
        table_name: tableName,
        summary: response,
        purpose: 'Data storage and retrieval',
        suggestions: [],
        common_queries: [],
        index_recommendations: [],
        relationship_suggestions: []
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        table_name: tableName,
        summary: '',
        suggestions: []
      };
    }
  }

  async suggestIndexes(tableName, columns, queryPatterns = null, dbType = 'postgresql') {
    try {
      const suggestions = [];

      for (const column of columns) {
        const colName = column.name || '';
        const colType = (column.type || '').toLowerCase();

        if (colName.toLowerCase().includes('id') && colName !== 'id') {
          suggestions.push({
            type: 'btree',
            columns: [colName],
            reason: `Foreign key column ${colName} would benefit from an index`
          });
        } else if (colType.includes('timestamp') || colType.includes('date') || colType.includes('datetime')) {
          suggestions.push({
            type: 'btree',
            columns: [colName],
            reason: `Date/time column ${colName} commonly used in WHERE clauses`
          });
        } else if (colName.toLowerCase().includes('email')) {
          suggestions.push({
            type: 'btree',
            columns: [colName],
            reason: `Email column ${colName} likely used for lookups`
          });
        }
      }

      return {
        success: true,
        table_name: tableName,
        index_suggestions: suggestions,
        performance_impact: 'medium'
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        index_suggestions: []
      };
    }
  }

  async analyzeRelationships(tables, dbType = 'postgresql') {
    try {
      const relationships = [];
      const tableNames = Object.keys(tables);

      for (const [tableName, columns] of Object.entries(tables)) {
        for (const column of columns) {
          const colName = (column.name || '').toLowerCase();

          if (colName.endsWith('_id') && colName !== 'id') {
            const referencedTable = colName.slice(0, -3);

            for (const otherTable of tableNames) {
              if (otherTable.toLowerCase() === referencedTable ||
                otherTable.toLowerCase() === referencedTable + 's') {
                relationships.push({
                  from_table: tableName,
                  from_column: column.name,
                  to_table: otherTable,
                  to_column: 'id',
                  relationship_type: 'many_to_one',
                  confidence: 'medium'
                });
              }
            }
          }
        }
      }

      return {
        success: true,
        relationships,
        total_suggestions: relationships.length
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        relationships: []
      };
    }
  }

  async analyzeSchema(schemaName, tables, dbType = 'postgresql') {
    try {
      const tableNames = Object.keys(tables);
      const tableCount = tableNames.length;

      if (tableCount === 0) {
        return {
          success: false,
          error: 'No tables found in schema',
          summary: '',
          design_patterns: [],
          relationships: [],
          optimization_suggestions: []
        };
      }

      // Analyze relationships across all tables
      const relationshipAnalysis = await this.analyzeRelationships(tables, dbType);
      const relationships = relationshipAnalysis.relationships || [];

      // Generate summary
      const summary = `Schema "${schemaName}" contains ${tableCount} table${tableCount !== 1 ? 's' : ''}. ` +
        `Detected ${relationships.length} relationship${relationships.length !== 1 ? 's' : ''} between tables.`;

      // Identify design patterns
      const design_patterns = [];

      // Check for common patterns
      const hasUserTable = tableNames.some(t => t.toLowerCase().includes('user'));
      const hasAuditFields = Object.values(tables).some(cols =>
        cols.some(c => ['created_at', 'updated_at', 'deleted_at'].includes(c.name?.toLowerCase()))
      );
      const hasJunctionTables = tableNames.some(t => t.includes('_'));

      if (hasUserTable) {
        design_patterns.push('User management pattern detected');
      }
      if (hasAuditFields) {
        design_patterns.push('Audit trail pattern with timestamp tracking');
      }
      if (hasJunctionTables) {
        design_patterns.push('Many-to-many relationships using junction tables');
      }
      if (relationships.length > tableCount) {
        design_patterns.push('Highly normalized schema with multiple relationships');
      }

      // Format relationships for display
      const formattedRelationships = relationships.map(rel =>
        `${rel.from_table}.${rel.from_column} → ${rel.to_table}.${rel.to_column}`
      );

      // Generate optimization suggestions
      const optimization_suggestions = [];

      // Check for missing indexes on foreign keys
      const foreignKeyColumns = relationships.map(r => ({
        table: r.from_table,
        column: r.from_column
      }));

      if (foreignKeyColumns.length > 0) {
        optimization_suggestions.push(
          `Consider adding indexes on foreign key columns for better join performance`
        );
      }

      // Check for tables without primary keys
      const tablesWithoutPK = Object.entries(tables).filter(([name, cols]) =>
        !cols.some(c => c.primary_key || c.name === 'id')
      );

      if (tablesWithoutPK.length > 0) {
        optimization_suggestions.push(
          `${tablesWithoutPK.length} table(s) may be missing primary keys: ${tablesWithoutPK.map(([name]) => name).join(', ')}`
        );
      }

      // Check for large tables without partitioning
      if (tableCount > 20) {
        optimization_suggestions.push(
          'Large schema detected - consider table partitioning for frequently accessed tables'
        );
      }

      return {
        success: true,
        schema_name: schemaName,
        summary,
        design_patterns,
        relationships: formattedRelationships.slice(0, 10), // Limit to 10 for display
        optimization_suggestions,
        table_count: tableCount,
        relationship_count: relationships.length
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        summary: '',
        design_patterns: [],
        relationships: [],
        optimization_suggestions: []
      };
    }
  }

  async generateCommonQueries(tableName, columns, dbType = 'postgresql') {
    try {
      const queries = [];

      queries.push({
        name: 'Select All Records',
        query: `SELECT * FROM ${tableName} LIMIT 10;`,
        description: 'Retrieve first 10 records from the table'
      });

      queries.push({
        name: 'Count Records',
        query: `SELECT COUNT(*) FROM ${tableName};`,
        description: 'Count total number of records'
      });

      for (const column of columns) {
        const colName = column.name || '';
        const colType = (column.type || '').toLowerCase();

        if (colType.includes('timestamp') || colType.includes('date')) {
          queries.push({
            name: `Recent Records by ${colName}`,
            query: `SELECT * FROM ${tableName} WHERE ${colName} >= NOW() - INTERVAL '7 days' ORDER BY ${colName} DESC;`,
            description: `Find records from the last 7 days based on ${colName}`
          });
        } else if (colName.toLowerCase().includes('email')) {
          queries.push({
            name: `Find by ${colName}`,
            query: `SELECT * FROM ${tableName} WHERE ${colName} = 'example@email.com';`,
            description: `Find records by ${colName}`
          });
        }
      }

      return {
        success: true,
        table_name: tableName,
        queries: queries.slice(0, 5),
        total_generated: queries.length
      };
    } catch (error) {
      return {
        success: false,
        error: error.message,
        queries: []
      };
    }
  }
}

let _schemaAnalyzer = null;

function getSchemaAnalyzer() {
  if (_schemaAnalyzer === null) {
    try {
      const { logger } = require('../../utils/logger');
      const { app } = require('electron');
      const path = require('path');

      // Load .env from correct location
      try {
        const dotenv = require('dotenv');
        const envPath = app.isPackaged
          ? path.join(process.resourcesPath, '.env')
          : path.join(__dirname, '../../../../../../.env');

        dotenv.config({ path: envPath });
        logger.info(`[Schema Analyzer] Loading .env from: ${envPath}`);
      } catch (envError) {
        logger.warn('[Schema Analyzer] Failed to load .env, using process env:', envError.message);
      }

      const accountId = process.env.CLOUDFLARE_ACCOUNT_ID;
      const apiToken = process.env.CLOUDFLARE_API_TOKEN;

      logger.info(`[Schema Analyzer] AI Config check - AccountID: ${accountId ? 'present' : 'missing'}, Token: ${apiToken ? 'present' : 'missing'}`);

      if (!accountId || !apiToken) {
        logger.warn('[Schema Analyzer] Cloudflare AI credentials not configured');
        return null;
      }

      _schemaAnalyzer = new SchemaAnalyzer(accountId, apiToken);
      logger.info('[Schema Analyzer] Schema analyzer initialized successfully');
    } catch (error) {
      const { logger } = require('../../utils/logger');
      logger.error('[Schema Analyzer] Failed to initialize schema analyzer:', error);
      return null;
    }
  }
  return _schemaAnalyzer;
}

module.exports = { SchemaAnalyzer, getSchemaAnalyzer };
```
